homeassistant/components/kraken/__init__.py

```diff
@@ -0,0 +1,163 @@
+"""The kraken integration."""
+import asyncio
+from datetime import timedelta
+import logging
+from typing import Dict
+
+import async_timeout
+import krakenex
+import pykrakenapi
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_SCAN_INTERVAL
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    CONF_TRACKED_ASSET_PAIRS,
+    DEFAULT_SCAN_INTERVAL,
+    DEFAULT_TRACKED_ASSET_PAIR,
+    DISPATCH_CONFIG_UPDATED,
+    DOMAIN,
+)
+from .utils import get_tradable_asset_pairs
+
+PLATFORMS = ["sensor"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the kraken component."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Set up kraken from a config entry."""
+    kraken_data = KrakenData(hass, config_entry)
+    await kraken_data.async_setup()
+    hass.data[DOMAIN] = kraken_data
+    config_entry.add_update_listener(async_options_updated)
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(config_entry, "sensor")
+    )
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        for unsub_listener in hass.data[DOMAIN].unsub_listeners:
+            unsub_listener()
+        hass.data.pop(DOMAIN)
+
+    return unload_ok
+
+
+class KrakenData:
+    """Define an object to hold kraken data."""
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+        """Initialize."""
+        self._hass = hass
+        self._config_entry = config_entry
+        self._api = pykrakenapi.KrakenAPI(krakenex.API(), retry=0, crl_sleep=0)
+        self.tradable_asset_pairs = None
+        self.coordinator = None
+        self.unsub_listeners = []
+
+    async def async_update(self) -> None:
+        """Get the latest data from the Kraken.com REST API.
+
+        All tradeable asset pairs are retrieved, not the tracked asset pairs
```

Comment:
Why would we care about removing of an asset pair that we don't track?

Reply:
Don't you use Config Entry Options to manage what assets to track?

Reply:
The way I implemented is as follows: Get a list all the available asset pairs and then request the latest data for the list of these asset pairs. When I get an error because Kraken delisted an asset I refresh the list.
This way I don't have to store 2 lists, one is the list of tracked asset pairs defined by the user and another filter list of known asset pairs which got delisted.
Furthermore I don't have to find out which asset got delisted. This is very complicated because it is not part of the error response by the api.